scheme/sevsnp/evidence_handler.go

```diff
@@ -0,0 +1,434 @@
+		return nil, err
+	}
+
+	err = refValComid.Valid()
+	if err != nil {
+		return nil, err
+	}
+
+	refValCorim := corim.UnsignedCorim{}
+	refValCorim.SetProfile("http://amd.com/2024/snp-corim-profile")
+	refValCorim.AddComid(refValComid)
+
+	refValJson, err := refValCorim.ToJSON()
+	if err != nil {
+		return nil, err
+	}
+
+	err = json.Unmarshal(refValJson, &claimsSet)
+	if err != nil {
+		return nil, err
+	}
+
+	return claimsSet, nil
+}
+
+// snpAttestationOptions parameter for verifying certificate chain
+func snpAttestationOptions() *verify.Options {
+	return &verify.Options{
+		Getter:              trust.DefaultHTTPSGetter(),
+		Now:                 time.Now(),
+		DisableCertFetching: true,
+	}
+}
+
+// readCert helper function to read a certificate from a blob
+func readCert(cert []byte) ([]byte, error) {
+	block, _ := pem.Decode(cert)
+	if block == nil || block.Type != "CERTIFICATE" {
+		return nil, fmt.Errorf("failed to read certificate")
+	}
+	return block.Bytes, nil
+}
+
+// compareTAs compares two X509 certificates for equality
+func compareTAs(provisionedArk []byte, evidenceArk []byte) (bool, error) {
+	pArk, err := readCert(provisionedArk)
+	if err != nil {
+		return false, err
+	}
+
+	pCert, err := x509.ParseCertificate(pArk)
+	if err != nil {
+		return false, err
+	}
+
+	eArk, err := readCert(evidenceArk)
+	if err != nil {
+		return false, err
+	}
+
+	eCert, err := x509.ParseCertificate(eArk)
+	if err != nil {
+		return false, err
+	}
+
+	return pCert.Equal(eCert), nil
+}
+
+// ValidateEvidenceIntegrity verifies that the ARK in the
+// evidence matches the provisioned ARK, confirms the
+// integrity of the certificate chain, and validates
+// the signature of the evidence.
+//
+// The "auxblob" in the evidence contains a certificate chain.
+// The Trust Anchor in this chain is AMD Root Key (ARK).
+func (o EvidenceHandler) ValidateEvidenceIntegrity(
+	token *proto.AttestationToken,
+	trustAnchors []string,
+	endorsementsStrings []string,
+) error {
+	var (
+		taEndorsement handler.Endorsement
+		avk           comid.KeyTriple
+		tsm           tokens.TSMReport
+	)
+
+	// Get the ARK TA
+	for i, t := range trustAnchors {
+		var endorsement handler.Endorsement
+
+		if err := json.Unmarshal([]byte(t), &endorsement); err != nil {
+			return fmt.Errorf("could not decode endorsement at index %d: %w", i, err)
+		}
+
+		if endorsement.Type == handler.EndorsementType_VERIFICATION_KEY {
+			taEndorsement = endorsement
+			break
+		}
+	}
+
+	if taEndorsement.Type != handler.EndorsementType_VERIFICATION_KEY {
+		return fmt.Errorf("trust anchors unavailable")
+	}
+
+	err := json.Unmarshal(taEndorsement.Attributes, &avk)
+	if err != nil {
+		return err
+	}
+
+	provisionedArk := avk.VerifKeys[0]
+
+	// Parse certificate chain in evidence (auxblob)
+	err = tsm.FromCBOR(token.Data)
+	if err != nil {
+		return err
+	}
+
+	protoReport, err := abi.ReportToProto(tsm.OutBlob)
+	if err != nil {
+		return err
+	}
+
+	ark, err := getARK(tsm.AuxBlob)
```

Comment:
ARK and ASK certs may not be in the AuxBlob and should always be provided by the user as their trust anchor in order to not be tricked with a fake SEV platform. I do see you have compareTAs, so the current logic is safe. I know I do the same thing is go-sev-guest, but I just filed a bug against myself to remove this.